functional/generics-type-classes.md

```diff
@@ -0,0 +1,268 @@
+boilerplate code for a `Vec` type (known as a `template`)  with two
+different types filled in.
+
+In Rust, a generic type parameter creates what is known in functional
+languages as a "type class constraint", and each different parameter filled in
+by an end user *actually changes the type*. In other words, `Vec<usize>` and
+`Vec<char>` *are two different types*, which are recognized as distinct by all
+parts of the type system.
+
+This is called **monomorphization**, where different types are created from
+**polymorphic** code.  This special behavior requires `impl` blocks to specify
+generic parameters: different values for the generic type cause different types,
+and different types can have different `impl` blocks.
+
+In object oriented languages, classes can inherit behavior from their parents.
+However, this allows the attachment of not only additional behavior to
+particular members of a type class, but extra behavior as well.
+
+The nearest equivalent is the runtime polymorphism in Javascript and Python,
+where new members can be added to objects willy-nilly by any constructor.
+Unlike those languages, however, all of Rust's additional methods can be type
+checked when they are used, because their generics are statically defined. That
+makes them more usable while remaining safe.
+
+An example follows.
+
+## Example
+
+Suppose you are designing a storage server for a series of lab machines.
+Because of the software involved, there are two different protocols you need
+to support: BOOTP (for PXE network boot), and NFS (for remote mount storage).
+
+Your goal is to have one program, written in Rust, which can handle both of
+them. It will have protocol handlers and listen for both kinds of requests. The
+main application logic will then allow a lab administrator to configure storage
+and security controls for the actual files.
+
+The requests from machines in the lab for files contain the same basic
+information, no matter what protocol they came from: an authentication method,
+and a file name to retrieve.  A straightforward implementation would look
+something like this:
+
+```rust,ignore
+struct FileDownloadRequest {
+    file_name: PathBuf,
+    authentication: Either<crate::nfs::AuthInfo, crate::bootp::AuthInfo>,
+}
+```
+
+This design might work well enough. But now suppose you needed to support
+adding metadata that was *protocol specific*. For example, with NFS, you
+wanted to determine what their mount point was in order to enforce additional
+security rules.
+
+The way the current struct is designed leaves the protocol decision until
+runtime. That means any method that applies to one protocol and not the other
+requires the programmer to do a runtime check.
+
+Here is how getting an NFS mount point would look:
+
+```rust,ignore
+struct FileDownloadRequest {
+    file_name: PathBuf,
+    authentication: Either<crate::nfs::AuthInfo, crate::bootp::AuthInfo>,
```

Comment:
`Either` is not a rust concept. This concept may not be familiar to a reader that comes from imperative languages. Or which approaches rust as their first programming language.
Should we have something like:

```
enum AuthInfo {
    Nfs(crate::nfs::AuthInfo),
    Bootp(crate::bootp::AuthInfo),
}
```
?

Reply:
`Either` is defined in [the `either` crate](http://docs.rs/either), which I think is easily discoverable. I will clarify that.

Reply:
Actually, the `enum` is better. I'll fix it.